### PR TITLE
Add sparse weight kernel

### PR DESCRIPTION
**Pull request by kalyc**
### Summary
Add sparse weight kernel

### Related Issues
Improve sparse keras benchmark performance

### PR Overview

- [ ] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [X] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n]

`PR has been tested on mnist_cnn & imdb_lstm example & end-to-end tests are passing`
We can consider merging PR to another branch until all sparse benchmark work is updated.

### CODE REVIEW

**Comment by kalyc** [comment]
Tested using H5 model saving - 

```
import time
from keras import Model
from keras.layers import Dense, Input
from keras.optimizers import SGD
from keras import backend as K
from keras.utils import multi_gpu_model
import keras as k
import mxnet as mx
from keras import models

def _validate_backend():
    if K.backend() != 'mxnet' and K.backend() != 'tensorflow':
        raise NotImplementedError('This benchmark script only supports MXNet and TensorFlow backend')

def run_benchmark(train_data, train_label, eval_data, eval_label, eval_iter, batch_size, epochs, gpus):
    _validate_backend()

    inputs = Input(batch_shape=(None, train_data.shape[1]), dtype='float32', sparse=True)
    predictions = Dense(units=1, activation='linear', kernel_initializer='normal')(inputs)
    model = Model(inputs=inputs, outputs=predictions)
    model.summary()

    sgd = SGD(lr=0.1, momentum=0.9)

    if gpus > 1:
        model = multi_gpu_model(model, gpus=gpus)

    model.compile(loss='mse', optimizer=sgd, metrics=['accuracy'])

    start = time.time()
    model.fit(train_data,
              train_label,
              epochs=epochs,
              batch_size=batch_size, verbose=1)

    data_names, data_shapes = models.save_mxnet_model(model=model, prefix='sparse_model')

    print("Keras Benchmark Results")
    print("Dataset: Synthetic Sparse Data")
    print("Backend: ", K.backend().capitalize())
    print("Batch Size: ", batch_size)
    if gpus >= 1:
        print("Benchmark run on %d GPU", gpus)
    else:
        print("Benchmark run on CPU")
    print("Total Time: ", time.time() - start)

    mse = model.evaluate(eval_data, eval_label, verbose=0, batch_size=batch_size)

    print("Achieved {0:.6f} validation MSE".format(mse[0]))
    print(model.evaluate(eval_data, eval_label, verbose=1, batch_size=batch_size))

    import tempfile
    _, fname = tempfile.mkstemp('.h5')
    models.save_model(model, fname)
    model = models.load_model(fname)

    print("Achieved {0:.6f} validation MSE".format(mse[0]))
    print(model.evaluate(eval_data, eval_label, verbose=1, batch_size=batch_size))

    sym, arg_params, aux_params = mx.model.load_checkpoint(prefix='sparse_model', epoch=0)

    mod = mx.mod.Module(symbol=sym,
                        data_names=data_names,
                        context=mx.cpu(),
                        label_names=None)

    mod.bind(for_training=False,
             data_shapes=data_shapes,
             label_shapes=mod._label_shapes)

    mod.set_params(arg_params, aux_params, allow_missing=True)

    metric = mx.metric.create('MSE')
    #metric = mx.metric.MSE()
    mse = mod.score(eval_iter, metric)

    print("Achieved {0:.6f} validation MSE".format(mse[0][1]))

```
In channels_first format on both Tensorflow & MXNet backend - 
Observing the same model performance - 
On MXNet - 

```
Keras Benchmark Results
Dataset: Synthetic Sparse Data
Backend:  Mxnet
Batch Size:  1024
Benchmark run on CPU
Total Time:  3.1575307846069336
Achieved 6143781.969072 validation MSE
99328/99328 [==============================] - 1s 5us/step
[6143781.969072165, 0.0]
Achieved 6143781.969072 validation MSE
/anaconda2/envs/mxnet/lib/python3.4/site-packages/mxnet/module/bucketing_module.py:408: UserWarning: Optimizer created manually outside Module but rescale_grad is not normalized to 1.0/batch_size/num_workers (1.0 vs. 0.0009765625). Is this intended?
  force_init=force_init)
99328/99328 [==============================] - 1s 5us/step
[6143781.969072165, 0.0]
```
On Tensorflow - 
```
Keras Benchmark Results
Dataset: Synthetic Sparse Data
Backend:  Tensorflow
Batch Size:  1024
Benchmark run on CPU
Total Time:  3.1681480407714844
Achieved 6287042.721649 validation MSE
99328/99328 [==============================] - 0s 4us/step
[6287042.721649485, 0.0]
Achieved 6287042.721649 validation MSE
99328/99328 [==============================] - 0s 5us/step
[6287042.721649485, 0.0]
```

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -360,7 +360,7 @@ def to_dense(tensor):
 name_scope = tf.name_scope
 
 
-def variable(value, dtype=None, name=None, constraint=None):
+def variable(value, dtype=None, name=None, constraint=None, sparse=False):
```
Comment (roywei) [inline keras/backend/tensorflow_backend.py]: Why do we need to change tensorflow backend? Please only modify mxnet_backend.py
If you what to pass sparse in Keras K.variable, add a check, so you won't break or change tensorflow backend
```
if backend == mxnet:
 K.variable(sparse = True)
else:
 K.variable()
```

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -369,6 +369,7 @@ def variable(value, dtype=None, name=None, constraint=None):
         name: Optional name string for the tensor.
         constraint: Optional projection function to be
             applied to the variable after an optimizer update.
+        sparse: Parameter used for MXNet backend to create sparse variable
```
Comment (roywei) [inline keras/backend/tensorflow_backend.py]: please remove

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -389,7 +390,7 @@ def variable(value, dtype=None, name=None, constraint=None):
     """
     if dtype is None:
         dtype = floatx()
-    if hasattr(value, 'tocoo'):
+    if hasattr(value, 'tocoo') or sparse:
```
Comment (roywei) [inline keras/backend/tensorflow_backend.py]: please remove

File: keras/engine/base_layer.py [hunk 4]
```diff
@@ -226,7 +226,8 @@ def add_weight(self,
                    initializer=None,
                    regularizer=None,
                    trainable=True,
-                   constraint=None):
+                   constraint=None,
```
Comment (roywei) [inline keras/engine/base_layer.py]: As mentioned, add backend check here and only add weight and pass sparse argument for mxnet backend.

File: keras/layers/core.py [hunk 5]
```diff
@@ -838,6 +838,7 @@ def __init__(self, units,
                  activity_regularizer=None,
                  kernel_constraint=None,
                  bias_constraint=None,
+                 sparse=False,
```
Comment (roywei) [inline keras/layers/core.py]: Why need to specify sparse during Dense initialization? How to use that?

File: tests/test_model_saving.py [hunk 6]
```diff
@@ -112,6 +113,30 @@ def test_functional_model_saving():
...
+    inputs = Input(shape=(3,), sparse=True)
+    x = Dense(2)(inputs)
+    outputs = Dense(3)(x)
+
+    model = Model(inputs, outputs)
+    model.compile(loss=losses.MSE,
+                  optimizer=optimizers.Adam(),
+                  metrics=[metrics.categorical_accuracy])
+    x = rand(1, 3, density=0.2, format='csr')
+    y = rand(1, 3, density=0.2, format='csr')
```
Comment (roywei) [inline tests/test_model_saving.py]: labels are not sparse, can we use normal dense labels

File: tests/test_model_saving.py [hunk 7]
```diff
@@ -149,6 +174,36 @@ def test_model_saving_to_pre_created_h5py_file():
...
```
Comment (roywei) [inline tests/test_model_saving.py]: use default format for label

File: tests/test_model_saving.py [hunk 8]
```diff
@@ -149,6 +174,36 @@ def test_model_saving_to_pre_created_h5py_file():
...
+    model.train_on_batch(x, y)
+
+    out = model.predict(x)
+    _, fname = tempfile.mkstemp('.h5')
+    with h5py.File(fname, mode='r+') as h5file:
```
Comment (roywei) [inline tests/test_model_saving.py]: what the difference between this section and `test_functional_model_saving_with_sparse`

File: tests/test_model_saving.py [hunk 9]
```diff
@@ -882,6 +978,43 @@ def test_sequential_lstm_mxnet_model_saving():
...
+    model.add(Embedding(max_features, 128, input_length=maxlen, sparse_grad=True))
+    model.add(LSTM(128))
+
+    model.compile(loss='binary_crossentropy',
+                  optimizer='adam',
+                  metrics=['accuracy'])
+
+    # Generate sparse random data
+    x = rand(1000, maxlen, density=0.2, format='csr')
+    y = rand(1000, 128, density=0.2, format='csr')
```
Comment (roywei) [inline tests/test_model_saving.py]: use dense labels

File: keras/backend/tensorflow_backend.py (hunk 1 above)
Comment (kalyc) [reply keras/backend/tensorflow_backend.py]: done

File: keras/backend/tensorflow_backend.py (hunk 2 above)
Comment (kalyc) [reply keras/backend/tensorflow_backend.py]: done

File: keras/backend/tensorflow_backend.py (hunk 3 above)
Comment (kalyc) [reply keras/backend/tensorflow_backend.py]: done

File: tests/test_model_saving.py (hunk 8 above)
Comment (kalyc) [reply tests/test_model_saving.py]: we are using a pre-created h5 file here`test_model_saving_to_pre_created_h5py...`

File: tests/test_model_saving.py (hunk 9 above)
Comment (kalyc) [reply tests/test_model_saving.py]: done

File: tests/test_model_saving.py (hunk 7 above)
Comment (kalyc) [reply tests/test_model_saving.py]: done

File: tests/test_model_saving.py (hunk 6 above)
Comment (kalyc) [reply tests/test_model_saving.py]: done

File: keras/backend/mxnet_backend.py [hunk 10]
```diff
@@ -263,6 +264,14 @@ def variable(value, dtype=None, name=None, constraint=None):
         ret._uses_learning_phase = False
         ret.bind(v)
         return ret
+    elif sparse:
```
[the remark this answers is not in the document]
Comment (kalyc) [reply keras/backend/mxnet_backend.py]: Yea, will update the documentation on sparse

File: keras/layers/core.py [hunk 11]
```diff
@@ -838,6 +838,7 @@ def __init__(self, units,
                  activity_regularizer=None,
                  kernel_constraint=None,
                  bias_constraint=None,
+                 sparse_weight=False,
```
Comment (roywei) [inline keras/layers/core.py]: Add explanation in doc string, note only used for mxnet

File: keras/engine/base_layer.py [hunk 12]
```diff
@@ -226,7 +226,8 @@ def add_weight(self,
...
+                   sparse_weight=False):
```
Comment (roywei) [inline keras/engine/base_layer.py]: Add explanation in doc string, note only used for mxnet

File: keras/layers/core.py [hunk 13]
```diff
@@ -863,7 +865,8 @@ def build(self, input_shape):
                                       initializer=self.kernel_initializer,
                                       name='kernel',
                                       regularizer=self.kernel_regularizer,
-                                      constraint=self.kernel_constraint)
+                                      constraint=self.kernel_constraint,
+                                      sparse_weight=self.sparse_weight)
```
Comment (roywei) [inline keras/layers/core.py]: what would happen here for tensorflow backend if Dense if initialized as Dense(sparse_weight=True)?
Shall we also add backend check here?

File: keras/layers/core.py (hunk 11 above)
Comment (kalyc) [reply keras/layers/core.py]: done

File: keras/engine/base_layer.py (hunk 12 above)
Comment (kalyc) [reply keras/engine/base_layer.py]: added

File: keras/backend/mxnet_backend.py [hunk 14]
```diff
@@ -263,6 +263,14 @@ def variable(value, dtype=None, name=None, constraint=None):
         ret._uses_learning_phase = False
         ret.bind(v)
         return ret
+    elif sparse_weight:
+        v = mx.ndarray.sparse.cast_storage(mx.nd.array(value), 'row_sparse')
```
Comment (roywei) [inline keras/backend/mxnet_backend.py]: Try creating row_sparse arrays using the row_sparse_array() API or mx.nd.array(value).tostype('row_sparse')
according to the doc:https://mxnet.incubator.apache.org/api/python/ndarray/sparse.html

Comment (kalyc) [reply keras/backend/mxnet_backend.py]: added
